Extract the packaging code into a service object

fredericboivin opened the pull request:
Logic related to building the ActiveShipping::Packages needed for the find rates is extracted into a service object.

The goal would be for it to be an extension point where different "packager" could be used. Code is the same for now, it's a straight extraction so no refactoring yet

Max weight is delegated to the calculator through Forwardable since max_weight_for_country is a protected method.

Next step would be to determine how to better handle the max_weight since it's determined on the calculator level and is currently riddled with bugs

Note : build is currently failing because USPS is returning 503

fredericboivin committed:
Extract packaging logic into PackageBuilder

Extract the packaging related code into a new service object.
Goal is for it to be an extension point which can support various
calculators related to different metrics. For now the code/logic
is exactly the same.

fredericboivin committed:
Extract common packaging methods to PackageHelper

Extract methods for quickly building inventory units and content items
into PackageHelper so it can be reused between spec

fredericboivin committed:
Access config values through a standard interface

Centralize the definition and use of the config values so it can
be easily updated and usage is the same everywhere

fredericboivin committed:
Simplify max_weight usage in package builder

Add comments to the method and simplify the logic. Compute it once
then reuse the result throughout the process

forkata commented on app/models/spree/calculator/weight/package_builder.rb:
```diff
@@ -0,0 +1,134 @@
+
+      attr_reader :shipping_calculator
+
+      def_instance_delegator :shipping_calculator, :max_weight_for_country
```
Is there any reason why we can't initialise the `PackageBuilder` with the `max_weight_for_county` so it doesn't have to be coupled with the calculator?

jhawthorn commented on app/models/spree/calculator/weight/package_builder.rb:
```diff
@@ -0,0 +1,134 @@
+module Spree
+  module Calculator::Weight
```
This isn't a Calculator, so it would be nice to pull it out of the Calculator namespace.

Maybe something like `SolidusActiveShipping::PackageBuilder` to avoid any possible confusion or conflicts.

fredericboivin committed:
Move the PackageBuilder to the Spree namespace

Delete the calculator::weight namespace and move the PackageBuilder
to the default Spree::

fredericboivin committed:
Simplify logic around computing weight from packages

Use the methods provided by ContentItem to avoid chaining associations
to get variant_weight + ContentItem. ContentItem also has a quantity
of one so we don't need the convoluted logic around fitting a max
quantity of item into a package.

fredericboivin committed:
Use the ProductPackage object, remove hardcoded unit

packages method was accessing the properties of ProductPackages
through an array index which is unreliable, opted to use the object
itself instead of an array of numbers and access the data through the
properties. Also remove the hardcoded imperial units

fredericboivin committed:
Distinguish Solidus packages from ActiveShipping packages

Rename variables so it's easier to understand what is coming from
Solidus and what is returned by ActiveShipping

fredericboivin committed:
Split generation of packages based on ProductPackages

Previously, product with associated product packages were computed twice
(once based on weight and added in the general packages) and once for
each product_packages. This commit aims to split the generation so
product are added once based on the existence of associated
product_packages

forkata commented on app/models/spree/calculator/weight/package_builder.rb:
```diff
@@ -15,11 +15,26 @@ def process(package)
       end
 
+      # Configuration 
+      def units
```
Are any of these going to be part of the public interface of the PackageBuilder? If not, i'd prefer if they were `private`.

fredericboivin replied:
To my knowledge, they should not be used elsewhere, so good catch

forkata commented on app/models/spree/calculator/weight/package_builder.rb:
```diff
@@ -132,11 +132,17 @@ def packages(package)
-        max_weight = max_weight_for_country(order.ship_address.country)
-        if max_weight == 0 && max_weight_per_package > 0
-          max_weight = max_weight_per_package
+        order_ship_address_country = order.ship_address.country
```
I am not sure if there is any benefit of pulling this out into a local, it is only referenced in one place in this method.

fredericboivin replied:
Fixed it

forkata commented on app/models/spree/package_builder.rb:
```diff
@@ -106,7 +106,7 @@ def packages(package)
 
       item_specific_packages.each do |package|
-        packages << ::ActiveShipping::Package.new(package.at(0), [package.at(1), package.at(2), package.at(3)], units: :imperial)
+        packages << ::ActiveShipping::Package.new(package.weight * multiplier, [package.length, package.width, package.height], units: units)
```
:+1: for removing the hard-coded imperial units. The default value for ActiveShipping is metric if I am not mistaken. Does this mean we need to initialize `Spree::ActiveShipping::Config[:units]` to `:imperial` by default to preserve the original functionality?

fredericboivin replied:
Everywhere else in the code, the unit is pulled from the config so I'd assume people who wants to use :imperial are setting the value themselves otherwise you'd get weird results

I'd also expect the behavior to be different because this whole part looks bugged to me considering that product with product packages are double-packed (once based on weight and once based on product_packages)

EDIT : Also the default preference is :imperial so it's okay

fredericboivin commented:
Last commit breaks the tests since the behavior is different, currently working on adding spec to the package_builder (that I will amend later) so we can kill/salvage the weight_limits_spec and kill the weight-related code in the base calculator

forkata commented on app/models/spree/package_builder.rb:
```diff
@@ -84,11 +84,24 @@ def convert_package_to_item_packages_array(solidus_package)
       active_shipping_packages = []
-      weights = convert_package_to_weights_array(solidus_package)
+
+      ci_without_boxes = []
```
I would prefer if we didn't introduce a new name for `product_packages` here as `boxes`. We also generally avoid abbreviating things in names. It isn't immediately clear what `ci` is when reading this code.

fredericboivin replied:
Makes sense, fixed it

forkata commented on app/models/spree/content_item_decorator.rb:
```diff
@@ -0,0 +1,9 @@
+module Spree
+  module Stock
+    ContentItem.class_eval do
+      def variant_has_product_packages?
```
This can be done through a 

```
delegate :has_product_packages?, to: :variant, prefix: true
```

I am not sure that I see the value of adding this class eval though.

fredericboivin replied:
Good point about the prefix, the goal is so we don't have to do something like content_item.variant.product.product_packages, it works like the backordered? or on_hand? scope on ContentItem to basically just filter the results

forkata commented on app/models/spree/product_decorator.rb:
```diff
@@ -3,4 +3,8 @@
   accepts_nested_attributes_for :product_packages, :allow_destroy => true, :reject_if => lambda { |pp| pp[:weight].blank? or Integer(pp[:weight]) < 1 }
+
+  def has_product_packages?
+  	product_packages.any?
```
Indentation looks off here.

forkata commented on app/models/spree/variant_decorator.rb:
```diff
@@ -0,0 +1,3 @@
+Spree::Variant.class_eval do
+	delegate :has_product_packages?, to: :product
```
Indentation looks off here.

fredericboivin replied:
Yeah, I'll run rubocop before I'm done, lots of small style issues lying around

fredericboivin commented:
@jhawthorn : I'd really appreciate if you could take a look and provide some feedback if you have the time (outside of style issues that I will rubocop and the stuff Chris already mentionned)

fredericboivin committed:
Remove weight_limit_spec

It's full of holes and it should be replace by package_builder_spec
instead.

fredericboivin committed:
Delete weight-related code from shipping calculators

The code was extracted to package_builder.rb

fredericboivin committed:
Add max_weight as PackageBuilder param

Before, the call to max_weight_for_country was delegated to the
calculator but the max_weight was computed by the package builder.
It is out of scope for the PackageBuilder since the max_weight
is determined at the calculator level. This commit remove the delegation
and make it a parameter of the process function. Also included : small
tweaks to the calculator to compute and send the parameters and changes
to the test suite.

fredericboivin committed:
Make config param private in the PackageBuilder

fredericboivin committed:
Run rubocop -a on edited files

Find indentation and various style issues

fredericboivin commented:
Should be ready for another round of review if build is green

fredericboivin commented:
Tests are broken in general now on master because of [Solidus PR#1138](https://github.com/solidusio/solidus/pull/1138)

Going to fix this in another PR

fredericboivin commented:
Will need to be rebased on #29

forkata commented:
#29 is merged so we can rebase this to get the tests :green_apple: again!

forkata commented:
This guy has been around for a while so I think it's time we take another look over it and merge :tada:

forkata commented:
@dangerdogz When you have a chance can you rebase this on top of master and push up so it can run specs again against new Solidus master and v2.0 branch.

fredericboivin commented:
Rebased
